File: app/src/main/java/com/faendir/lightning_launcher/multitool/fastadapter/ExpandableItem.java
package com.faendir.lightning_launcher.multitool.fastadapter;

import android.content.Context;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.InsetDrawable;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.support.v7.widget.RecyclerView;
import android.util.DisplayMetrics;
import android.view.View;
import android.widget.TextView;
import com.faendir.lightning_launcher.multitool.R;
import com.mikepenz.fastadapter.IItem;
import com.mikepenz.fastadapter.expandable.items.ModelAbstractExpandableItem;
import com.mikepenz.fastadapter_extensions.swipe.ISwipeable;

import java.util.List;

/**
 * Created by Lukas on 23.08.2015.
 * An item in the ExpandableListView
 */
public class ExpandableItem<T extends Model> extends ModelAbstractExpandableItem<T, ExpandableItem<T>, ExpandableItem.ViewHolder, ExpandableItem<T>> implements ISwipeable<ExpandableItem<T>, IItem> {
    private final int size;
    private boolean swipeable;
    private Runnable swipedAction;

    public ExpandableItem(T item, int size) {
        super(item);
        swipeable = item instanceof DeletableModel;
        this.size = size;
        if (item instanceof ClickAwareModel) {
            withOnItemClickListener((v, adapter, item1, position) -> {
                ((ClickAwareModel) item).onClick();
                return true;
            });
        }
    }

    @NonNull
    @Override
    public ViewHolder getViewHolder(@NonNull View v) {
        return new ViewHolder(v);
    }

    @Override
    public int getType() {
        return R.id.txt;
    }

    @Override
    public int getLayoutRes() {
        return R.layout.list_item_app;
    }

    @Override
    public void bindView(ViewHolder holder, List<Object> payloads) {
        super.bindView(holder, payloads);
        Context context = holder.text.getContext();
        Drawable icon;
        String text;
        int background;
        if (swipedAction == null) {
            text = getModel().getName();
            icon = DrawableCompat.wrap(getModel().getIcon(context));
            DrawableCompat.setTint(icon, getModel().getTintColor());
            DrawableCompat.setTintMode(icon, PorterDuff.Mode.MULTIPLY);
            background = isSelected() ? context.getResources().getColor(R.color.accent) : Color.TRANSPARENT;
        } else {
            text = ((DeletableModel) getModel()).getUndoText(context);
            icon = context.getResources().getDrawable(R.drawable.ic_delete_white);
            Drawable ref = getModel().getIcon(context);
            float diffX = ((float) ref.getIntrinsicWidth() - icon.getIntrinsicWidth()) / 2;
            float diffY = ((float) ref.getIntrinsicHeight() - icon.getIntrinsicHeight()) / 2;
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                icon = new InsetDrawable(icon, diffX, diffY, diffX, diffY);
            } else {
                icon = new InsetDrawable(icon, Math.round(diffX), Math.round(diffY), Math.round(diffX), Math.round(diffY));
            }
            background = Color.RED;
        }
        holder.text.setText(text);
        icon.setBounds(0, 0, size * icon.getIntrinsicWidth() / icon.getIntrinsicHeight(), size);
        holder.text.setCompoundDrawables(icon.mutate(), null, null, null);
        holder.text.setBackgroundColor(background);
        holder.applyInset(10 * getLevel());
    }

    private int getLevel() {
        int level;
        ExpandableItem<?> parent = getParent();
        for (level = 0; parent != null; level++) {
            parent = parent.getParent();
        }
        return level;
    }

    @Override
    public String toString() {
        return "ExpandableItem{" + getModel().toString() + "}";
    }

    @Override
    public boolean isSwipeable() {
        return swipeable;
    }

    @Override
    public ExpandableItem<T> withIsSwipeable(boolean swipeable) {
        this.swipeable = swipeable;
        return this;
    }

    public void setSwipedAction(Runnable action) {
        this.swipedAction = action;
        withOnItemPreClickListener((v, adapter, item, position) -> {
            if (this.swipedAction != null) {
                swipedAction.run();
                return true;
            }
            return false;
        });
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        public final TextView text;
        private final int basePadding;

        ViewHolder(View itemView) {
            super(itemView);
            text = itemView.findViewById(R.id.txt);
            basePadding = itemView.getPaddingLeft();
        }

        void applyInset(int insetDp) {
            DisplayMetrics displayMetrics = itemView.getContext().getResources().getDisplayMetrics();
            int inset = Math.round(insetDp * displayMetrics.density);
            itemView.setPadding(basePadding + inset, itemView.getPaddingTop(), itemView.getPaddingRight(), itemView.getPaddingBottom());
        }
    }
}

File: app/src/main/java/com/faendir/lightning_launcher/multitool/fastadapter/ExpandableItem.kt
package com.faendir.lightning_launcher.multitool.fastadapter

import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.drawable.Drawable
import android.graphics.drawable.InsetDrawable
import android.os.Build
import android.view.View
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.DrawableCompat
import androidx.recyclerview.widget.RecyclerView
import com.faendir.lightning_launcher.multitool.R
import com.mikepenz.fastadapter.IItem
import com.mikepenz.fastadapter.expandable.items.ModelAbstractExpandableItem
import com.mikepenz.fastadapter_extensions.swipe.ISwipeable

/**
 * Created by Lukas on 23.08.2015.
 * An item in the ExpandableListView
 */
class ExpandableItem<T : Model>(item: T, private val size: Int) : ModelAbstractExpandableItem<T, ExpandableItem<T>, ExpandableItem.ViewHolder, ExpandableItem<T>>(item), ISwipeable<ExpandableItem<T>, IItem<*, *>> {
    private var swipeable: Boolean = item is DeletableModel
    private var swipedAction: (()->Unit?)? = null

    private val level: Int
        get() {
            var level = 0
            var parent: ExpandableItem<*>? = parent
            while (parent != null) {
                parent = parent.parent
                level++
            }
            return level
        }

    init {
        if (item is ClickAwareModel) {
            withOnItemClickListener { _, _, _, _ ->
                (item as ClickAwareModel).onClick()
                true
            }
        }
    }

    override fun getViewHolder(v: View): ViewHolder = ViewHolder(v)

    override fun getType(): Int = R.id.txt

    override fun getLayoutRes(): Int = R.layout.list_item_app

    override fun bindView(holder: ViewHolder, payloads: MutableList<Any>) {
        super.bindView(holder, payloads)
        val context = holder.text.context
        var icon: Drawable
        val text: String
        val background: Int
        if (swipedAction == null) {
            text = model.name
            icon = DrawableCompat.wrap(model.getIcon(context))
            DrawableCompat.setTint(icon, model.tintColor)
            DrawableCompat.setTintMode(icon, PorterDuff.Mode.MULTIPLY)
            background = if (isSelected) ContextCompat.getColor(context, R.color.accent) else Color.TRANSPARENT
        } else {
            text = (model as DeletableModel).getUndoText(context)
            icon = ContextCompat.getDrawable(context, R.drawable.ic_delete_white)!!
            val ref = model.getIcon(context)
            val diffX = (ref.intrinsicWidth.toFloat() - icon.intrinsicWidth) / 2
            val diffY = (ref.intrinsicHeight.toFloat() - icon.intrinsicHeight) / 2
            icon = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                InsetDrawable(icon, diffX, diffY, diffX, diffY)
            } else {
                InsetDrawable(icon, Math.round(diffX), Math.round(diffY), Math.round(diffX), Math.round(diffY))
            }
            background = Color.RED
        }
        holder.text.text = text
        icon.setBounds(0, 0, size * icon.intrinsicWidth / icon.intrinsicHeight, size)
        holder.text.setCompoundDrawables(icon.mutate(), null, null, null)
        holder.text.setBackgroundColor(background)
        holder.applyInset(10 * level)
    }

    override fun toString(): String = "ExpandableItem{$model}"

    override fun isSwipeable(): Boolean = swipeable

    override fun withIsSwipeable(swipeable: Boolean): ExpandableItem<T> {
        this.swipeable = swipeable
        return this
    }

    fun setSwipedAction(action: (()->Unit?)?) {
        this.swipedAction = action
        withOnItemPreClickListener { _, _, _, _ -> (this.swipedAction != null).also { if (it) swipedAction?.invoke() } }
    }

    class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val text: TextView = itemView.findViewById(R.id.txt)
        private val basePadding: Int = itemView.paddingLeft

        internal fun applyInset(insetDp: Int) {
            val displayMetrics = itemView.context.resources.displayMetrics
            val inset = Math.round(insetDp * displayMetrics.density)
            itemView.setPadding(basePadding + inset, itemView.paddingTop, itemView.paddingRight, itemView.paddingBottom)
        }
    }
}
